### PR TITLE
Add more diagnostics to heartbeat timing

### PR DESCRIPTION
**Pull request by pcd1193182**

### CODE REVIEW

File: cmd/zfs_object_agent/zettaobject/src/heartbeat.rs [hunk 1]
```diff
@@ -167,6 +177,16 @@ pub async fn start_heartbeat(object_access: ObjectAccess, id: Uuid) -> Heartbeat
...
+                    .unwrap();
+                if since > *HEARTBEAT_INTERVAL {
+                    trace!("Heartbeat locking slightly over: {:?}", since);
+                } else if since > 2 * *HEARTBEAT_INTERVAL {
+                    debug!("Heartbeat locking significantly over: {:?}", since);
+                }
+            }
             let heartbeat = HeartbeatPhys {
                 timestamp: SystemTime::now(),
                 hostname: hostname::get().unwrap().into_string().unwrap(),
```
Comment (ahrens) [inline cmd/zfs_object_agent/zettaobject/src/heartbeat.rs:192]: could you also time the hostname::get()?

**Review by ahrens — COMMENTED** [review]
Another potential issue I was thinking about (but doesn't seem to be in play in the log I obtained) is that we are comparing the SystemTime's to determine how much time elapsed since the last heartbeat.  However, SystemTime is just informative and (I think) we designed it so that nobody needs to care about clocks being in sync.  The SystemTime could change (jump forward or backward), which would mess up our idea of how much time elapsed.  We would be better off using Instant, which should move forward at a constant rate.

File: cmd/zfs_object_agent/zettaobject/src/heartbeat.rs (hunk 1 above)
Comment (pcd1193182) [reply cmd/zfs_object_agent/zettaobject/src/heartbeat.rs:192]: I figured I would leave this one as implicitly timed; the step immediately after it is the `put`, which you've already shown has timestamps associated with it. I can add an explicit message instead

**Comment by pcd1193182** [comment]
The problem with using Instant is that it can't be serialized into a Heartbeat; this is explicitly why we used SystemTime, because we didn't think we'd care much if the heartbeat time skipped around a little bit. I guess we do care for lease timeouts. We could use the Instant to measure the elapsed time internally, but serialize a SystemTime to the heartbeat? We'll still have skip issues if the clock changes while we're restarting the agent, but I think that can't be helped.

**Comment by ahrens** [comment]
> We could use the Instant to measure the elapsed time internally, but serialize a SystemTime to the heartbeat?

Yeah I think that would make sense.  We could even replace the SystemTime in the heartbeatphys with a GUID (although it might make sense to keep the SystemTime additionally, to display to the user).

> We'll still have skip issues if the clock changes while we're restarting the agent, but I think that can't be helped.

I think that in the current design, the places that the SystemTime in the heartbeat is used are:
* in lease_timed_out(), which we're taling about changing
* in try_claim(), where we check if it's the same as what we saw before.  In other words, the SystemTime is used like a GUID to distinguish instances of the HeartbeatPhys, to make sure that it doesn't change while we are waiting for an existing lease to timeout.

So I don't think that there would be "skip issues" when restarting the agent.

File: cmd/zfs_object_agent/zettaobject/src/heartbeat.rs (hunk 1 above)
Comment (ahrens) [reply cmd/zfs_object_agent/zettaobject/src/heartbeat.rs:192]: OK, so if it isn't either of the 2 new "significantly over" messages then it must be the hostname::get(), and then we'll move it outside of the loop.

**Review by ahrens — APPROVED** [review]

**Pull request merged**